Getting started update milestone 3 (#191)
stratum-mining/stratumprotocol.org: src/getting-started.md

# Getting Started with running SRI - Stratum v2 reference implementation

This document aims to assist users in deploying the SRI software stack efficiently. Stratum v2 as a protocol is flexible, and allows users to run multiple configurations. In the getting started guide, we’re focusing on running [roles](https://stratumprotocol.org/docs/#roles) which allow miners to construct their own block templates. 

![SRI ](/assets/Sri-AllRoles.png)

## Prerequisites

Rust installed on your machine. If it's not:

```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```
There are two primary ways to use all roles:  

- Run certain roles and connecting to our community-hosted roles. (The easiest way to test SRI)
- Run all roles (Recommended way)

## I Getting Started - The easiest way

**Objective**: Give Stratum v2 a try, mine blocks on the testnet with minimal setup.

- Community-hosted roles used: SV2 Pool, testnet Template Provider, and Job Declarator Server.
- Self-hosted roles to run: Job Declarator Client (JDC) and Translator Proxy.

### Step 1: Setup and Configuration

#### Clone the repository

```bash
git clone https://github.com/stratum-mining/stratum.git

> git checkout master

Author: Wow didn't expect that @Fi3 failed under political correctness pressure! Thanks for catching this one.

```
#### Run Job Declarator Client (JDC)
```bash
cd roles/jd-client/config-examples/

> cargo run -- -c proxy-config-hosted-example.toml

Reviewer: ```suggestion
cargo run -- -c jdc-config-hosted-example.toml
```

```
#### Run Translator Proxy

```bash
cd roles/translator/config-examples/

> cargo run -- -c proxy-config-local-jdc-example.toml

Reviewer: ```suggestion
cargo run -- -c tproxy-config-local-jdc-example.toml
```

```
#### Connect mining devices

Connect mining device - Translator Proxy will be running on port `34255`, so you will need just to edit your mining device(s) configuration, adding the line:

> `stratum+tcp://<tProxy ip>:34255`

Reviewer: ```suggestion
`stratum+tcp://<tProxy ip>:34255`

#### CPU Miner
If you don't have a physical miner, you can do tests with CPUMiner.

Setup the correct CPUMiner for your OS:
- You can download the binary directly from [here](https://sourceforge.net/projects/cpuminer/files/);
- Or compile it from [https://github.com/pooler/cpuminer](https://github.com/pooler/cpuminer)

On the CPUMiner directory:
 
`./minerd -a sha256d -o stratum+tcp://localhost:34255 -q -D -P`

#### Adjust proxy-config (optional)

> Depending on mining device you do run, you may have to adjust proxy-config file in order to adjust the min_individual_miner_hashrate and channel_nominal_hashrate parameters accordingly

Reviewer: ```suggestion
Depending on mining device you do run, you may have to adjust `tproxy-config-local-jdc-example.toml` file in order to adjust the `min_individual_miner_hashrate` and `channel_nominal_hashrate` parameters accordingly
```

> You’re all set up, proceed to the `Final Step` section.

Reviewer: ```suggestion
```

this sentence doesn't make a lot of sense

## II Getting Started - Running all roles 

**Objective**: Establish a complete, self-hosted Stratum V2 environment for comprehensive testing and deeper understanding of all components.

### Step 1: Setup and Configuration

#### Run Template Provider

Clone template provider repository:

```bash

> git clone https://github.com/Sjors/bitcoin.git

Reviewer: ```suggestion
git clone https://github.com/Sjors/bitcoin.git
cd bitcoin
git checkout 2023/11/sv2-poll
```

> ```bash
> git clone https://github.com/Sjors/bitcoin.git
> ```

Reviewer: ```suggestion
```bash
git clone https://github.com/Sjors/bitcoin.git
```

Make sure you check out the `2023/11/sv2-poll` branch:
```
cd bitcoin
git checkout 2023/11/sv2-poll
```
```

Next, compile the template provider:

```bash
make clean
./autogen.sh && ./configure 
make -j 10
```

> Edit the `bitcoin.conf`` file by adding:

Reviewer: ```suggestion
Edit the `bitcoin.conf` file by adding:
```

Run the Template provider:

```bash
./src/bitcoind -sv2 -sv2port=8442 -debug=sv2 
```
Optional paremeters:

There are optional parameters which can be used to better manage the Template Provider:
- `sv2interval` - sets how often a new template is built (default is 30s)
- `sv2feedelta` - defines the delta fees to reach before sending new templates to downstreams (default is 1000 sats)
- `loglevel=sv2:trace` to get more detailed debugging

For example: 

```bash
./src/bitcoind -sv2 -sv2port=8442 -sv2interval=2 -sv2feedelta=1000 -debug=sv2 -loglevel=sv2:trace
```
This way new templates are constructed every 2 seconds (taking the most profitable txs from the mempool) and they are sent downstream if new fees collected are more than 1000 sats. 

#### Clone the repository

```bash
git clone https://github.com/stratum-mining/stratum.git
git checkout master
```
#### Run the SV2 Pool

```bash
cd roles/pool/config-examples
cargo run -- -c pool-config-local-tp-example.toml
```
#### Run the Job Declarator Server (JDS)

```bash
cd roles/jd-server/config-examples
cargo run -- -c jds-config-local-example.toml
```
#### Run Job Declarator Client (JDC)

```bash
cd roles/jd-client/config-examples/

> cargo run -- -c proxy-config-local-example.toml

Reviewer: ```suggestion
cargo run -- -c jdc-config-local-example.toml
```

```
#### Run Translator Proxy

```bash
cd roles/translator/config-examples/

> cargo run -- -c proxy-config-local-jdc-example.toml

Reviewer: ```suggestion
cargo run -- -c tproxy-config-local-jdc-example.toml
```

```
#### Connect mining devices

Connect mining device - Translator Proxy will be running on port `34255`, so you will need just to edit your mining device(s) configuration, adding the line:

`stratum+tcp://<tProxy ip>:34255`

> 

Reviewer: ```suggestion
#### CPU Miner
If you don't have a physical miner, you can do tests with CPUMiner.

Setup the correct CPUMiner for your OS:
- You can download the binary directly from [here](https://sourceforge.net/projects/cpuminer/files/);
- Or compile it from [https://github.com/pooler/cpuminer](https://github.com/pooler/cpuminer)

On the CPUMiner directory:
 
`./minerd -a sha256d -o stratum+tcp://localhost:34255 -q -D -P`

# Getting Started with running SRI - Stratum v2 reference implementation

This document aims to assist users in deploying the SRI software stack efficiently. Stratum v2 as a protocol is flexible, and allows users to run multiple configurations. In the getting started guide, we’re focusing on running [roles](https://stratumprotocol.org/docs/#roles) which allow miners to construct their own block templates. 

![SRI ](/assets/Sri-AllRoles.png)

## Prerequisites

Rust installed on your machine. If it's not:

```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```
There are two primary ways to use all roles:  

- Run certain roles and connecting to our community-hosted roles. (The easiest way to test SRI)
- Run all roles (Recommended way)

## I Getting Started - The easiest way

**Objective**: Give Stratum v2 a try, mine blocks on the testnet with minimal setup.

> - Community-hosted roles used: SV2 Pool, testnet Template Provider, and Job Declarator Server.

Author: I think on the website we use SV2, so that's why, did we have a discussion around this, what's a preferred way?

> - Community-hosted roles used: SV2 Pool, testnet Template Provider, and Job Declarator Server.

Reviewer 2: > I think on the website we use SV2, so that's why, did we have a discussion around this, what's a preferred way?

It seems to me that we are mixing it also on website. 
In [docs](https://stratumprotocol.org/docs/) we use `Sv2` and `Sv1`, everywhere else is `SV2` and `SV1`. 
I don't have an opinion on this, the important thing is to have it in a uniformed way imo

> - Community-hosted roles used: SV2 Pool, testnet Template Provider, and Job Declarator Server.

Author: I think we should file a separate issue for it and fix it once there's consensus, Sv2 sounds good to me, but wouldn't block the progress of this PR over the linguistic.

> - Community-hosted roles used: SV2 Pool, testnet Template Provider, and Job Declarator Server.

Reviewer 2: > I think we should file a separate issue for it and fix it once there's consensus, Sv2 sounds good to me, but wouldn't block the progress of this PR over the linguistic.

I agree with you.
Let's merge it 👍

- Self-hosted roles to run: Job Declarator Client (JDC) and Translator Proxy.

### Step 1: Setup and Configuration

#### Clone the repository

```bash
git clone https://github.com/stratum-mining/stratum.git
git checkout main
```
#### Run Job Declarator Client (JDC)
```bash
cd roles/jd-client/config-examples/
cargo run -- -c jdc-config-hosted-example.toml
```
#### Run Translator Proxy

```bash
cd roles/translator/config-examples/
cargo run -- -c tproxy-config-local-jdc-example.toml
```

> #### Connect mining devices

Author: Resolved in https://github.com/stratum-mining/stratumprotocol.org/pull/191/commits/f7a1ab5edd1a1deb3f926f6fcc898398a353d056

- Or compile it from [https://github.com/pooler/cpuminer](https://github.com/pooler/cpuminer)

On the CPUMiner directory:
 
`./minerd -a sha256d -o stratum+tcp://localhost:34255 -q -D -P`
## III Final Step: monitoring for blocks

Once set up, monitor the machines and role logs for any valid blocks found.

Following pictures detail the logs from tProxy, JDC, and Pool when a block is found.

Pool logs

![SRI Pool](/assets/SV2Pool-logs.png)

Job Declarator Client (JDC) logs

![SRI JDC](/assets/JobDeclaratorClient-JDC-logs.png)

Translation Proxy tProxy logs

![SRI tProxy](/assets/TranslationProxy-tProxy-logs.png) 

### Monitor for mined blocks on Testnet

If you didn’t change anything in the configuration files present in every role subdirectory, the valid block mined will be recognized by the string `Stratum v2 SRI Pool (0x5374726174756d2076322053524920506f6f6c)` inserted into the scriptSig of the coinbase tx input. 

A way to look for it is to check the address `tb1qa0sm0hxzj0x25rh8gw5xlzwlsfvvyz8u96w3p8` since the default coinbase output is configured to be a P2WPKH built with a public key = `036adc3bdf21e6f9a0f0fb0066bf517e5b7909ed1563d6958a10993849a7554075`

> You can performa search on one of the following block explorers:

Author: Resolved in https://github.com/stratum-mining/stratumprotocol.org/pull/191/commits/f7a1ab5edd1a1deb3f926f6fcc898398a353d056

> - https://mempool.space/testnet 

Reviewer 2: Hypertext is not working.
Maybe it is intended?

Author: https://github.com/stratum-mining/stratumprotocol.org/pull/191/commits/f7a1ab5edd1a1deb3f926f6fcc898398a353d056

> - https://blockstream.info/testnet/ 

Reviewer 2: Same here

Author: Resolved in https://github.com/stratum-mining/stratumprotocol.org/pull/191/commits/f7a1ab5edd1a1deb3f926f6fcc898398a353d056

## IV Customizing configuration

### Customize the “pool_signature” string
To customize the string which is inserted into the `scriptSig` of the coinbase tx input, you can change it by editing the `pool_signature` field present in:

> - roles/pool/pool-config.toml

Reviewer 2: The config referenced here does not exist anymore.
I would suggest to put `roles/pool/config-examples/pool-config-local-tp-example.toml`

Author: Resolved in https://github.com/stratum-mining/stratumprotocol.org/pull/191/commits/f7a1ab5edd1a1deb3f926f6fcc898398a353d056

> - roles/jd-client/jd-client-config.toml

Reviewer 2: Same here: `roles/jd-client/config-examples/jdc-config-local-example.toml`

Double check that the strings are equal and that you save your changes in both configuration files!

### Customize the coinbase tx output script

> To customize the coinbase tx output script, edit the `coinbase_output` field present in:

Reviewer 2: Personal suggestion: "To customize the coinbase tx output script, you need to run **every role locally**. You can customize it by editing the ...."

> - roles/pool/pool-config.toml

Reviewer 2: Same as above: `roles/pool/config-examples/pool-config-local-tp-example.toml`

Author: Resolved in https://github.com/stratum-mining/stratumprotocol.org/pull/191/commits/f7a1ab5edd1a1deb3f926f6fcc898398a353d056

> - roles/jd-server/jd-server-config.toml

Reviewer 2: Same here: `roles/jd-server/config-examples/jds-config-local-example.toml`

Author: Resolved in https://github.com/stratum-mining/stratumprotocol.org/pull/191/commits/f7a1ab5edd1a1deb3f926f6fcc898398a353d056

You can use any of the following script types:  P2PK,P2PKH, P2WPKH, P2SH, P2WSH, P2TR. 

> In case a public key is required, as explained in the config files, you can start by creating a testnet wallet on mobile, using Green wallet, or a desktop one, using Electrum wallet, and extract the extended public key they provide. At this point, you can derive the child public key to use in the configuration files, using the script we developed as SRI. Please have a look at https://github.com/stratum-mining/stratum/tree/dev/utils/bip32-key-derivation.

Reviewer 2: Hypertext is not working.
Maybe it is intended?
